### PR TITLE
fix(4895): handle measurement missing, null bytes, and `=` in measurement names

### LINKED ISSUE #4895: Schema validation error with null byte (0x00)
The routers are reporting the following:

```
level=error msg="schema validation failed" namespace=<snip> error="unhandled sqlx error: error returned from database: invalid byte sequence for encoding \"UTF8\": 0x00" target="router::dml_handlers::schema_validation" location="router/src/dml_handlers/schema_validation.rs:213" time=1655459035017563450
```

This has been repeated relatively frequently (~30s interval).

All occurrences I've observed have been for the same namespace, suggesting it is caused by specific user input.

### PR DESCRIPTION
**Pull request by wiedld**
Closes #4895 

Problems is parsing measurement from line protocol:
* reported issue with null bytes `\0`
* found issue with missing measurement, which resulted in either:
   * obtuse error messages, due to parsing errors then occurring further down the lp parsed line (a.k.a. later tokens errored) 
       * => documented in test case
   * or, a field_set or tag_set was erroneously recognized as a valid measurement => with later errors in namespace lookup 
      * => matches influxdb golang parser. Make contract explicit via tests.

---

* test: add tests for the desired contract for parsing measurements from line protocol (e251c3a86f8d3dbcb34deae08c6c41ef83fe05b6)

* fix: restrict null chars in measurement (b1818c801849682e8e2284a11ffbae103794783f)

* fix: prevent a tag_set or field_set being recognized as a measurement (74763352f10f2245349c448f2d11df16b35ef42)
    ```
    * the measurement parser was greedy, and in the absence of a measurement would recognize
     the next item (tag_set or field_set) as a measurement. 
    * with this greedy recognition we sometimes got an error later in parsing, whereas in other items
     it would not error and continue with the incorrect measurement name (to later fail in the namespace lookup). 
    * we could have fixed a lookahead peek, without consuming, and see if it's a valid measurement 
    (not tag_set or field_set). But that would halt parsing anyways with an error at this point. 
    So instead choose a normal consuming take parse and then error on failed verify. 
    ```

* chore: make an explicit Measurement type (9d3194a1fd00f1db00bb1450d863b72b6207d011)

* test: fix existing test, since a WITH_EQ is no longer a valid measurement (9243ccfb8756c56e78e85389b87d26b76af57291)

Additional commits per code review:

*  test: additional lp parsing tests, to make explicit where nulls are accepted (a559cbf52e7e9c60b75c4bd18e88a35328750484)

* Revert "test: fix existing test, since a WITH_EQ is no longer a valid measurement" (d6667c0ccbe5353d2b6fcc82d6fbb94ae76da959)

* refactor: have iox lp parser match influxdb contract, for acceptance of eq in measurements (3718a78d15be1bc7815d7a3505d851033a960591)
   ```
    * update contracts to match
    * create end_to_end test to confirm same write-then-read (the quoted SQL read) as influxdb
   ```

## checklist 
- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: influxdb_line_protocol/src/builder.rs [hunk 1]
```diff
@@ -418,26 +418,22 @@ mod tests {
             .field("dummy", true)
             .close_line()
             // line 7
-            .measurement(WITH_EQ)
```
Comment (wiedld) [inline influxdb_line_protocol/src/builder.rs:421]: No longer permitted.

File: influxdb_line_protocol/src/lib.rs [hunk 2]
```diff
@@ -640,16 +645,22 @@ fn series(i: &str) -> IResult<&str, Series<'_>> {
     )(i)
 }
 
-fn measurement(i: &str) -> IResult<&str, EscapedStr<'_>> {
-    let normal_char = take_while1(|c| !is_whitespace_boundary_char(c) && c != ',' && c != '\\');
+fn measurement(i: &str) -> IResult<&str, Measurement<'_>, Error> {
+    let normal_char = take_while1(|c| {
+        !is_whitespace_boundary_char(c) && !is_null_char(c) && c != ',' && c != '\\'
+    });
+    let verified = verify(normal_char, |str: &str| str.find('=').is_none());
```
Comment (wiedld) [inline influxdb_line_protocol/src/lib.rs:652]: This probably could be done with a field_set or tag_set pattern recognition. But i couldn't figure out the proper nom combination. with the verify.

File: influxdb_line_protocol/src/lib.rs [hunk 3]
```diff
@@ -1246,6 +1261,32 @@ mod test {
...
+        let vals = parse(input);
+        assert!(matches!(vals, Err(Error::MeasurementValueInvalid)));
+
+        let input = ",tag1=1,tag2=2 value=1 123";
+        let vals = parse(input);
+        assert!(matches!(vals, Err(Error::MeasurementValueInvalid)));
+
+        let input = "tag1=1,tag2=2 value=1 123";
+        let vals = parse(input);
+        assert!(matches!(vals, Err(Error::MeasurementValueInvalid)));
```
Comment (wiedld) [inline influxdb_line_protocol/src/lib.rs:1276]: Previously, this was returning `Measurement: SingleSlice("tag1=1"`:
```
Ok([ParsedLine { series: Series { raw_input: "tag1=1,tag2=2", measurement: SingleSlice("tag1=1"), tag_set: Some([(SingleSlice("tag2"), SingleSlice("2"))]) }, field_set: [(SingleSlice("value"), F64(1.0))], timestamp: Some(123) }])
```

File: influxdb_line_protocol/src/lib.rs [hunk 4]
```diff
@@ -1246,6 +1261,32 @@ mod test {
...
+
+    #[test]
+    fn parse_null_measurement() {
+        let input = "\0 field=1 1234";
+        let vals = parse(input);
+        assert!(matches!(vals, Err(Error::MeasurementValueInvalid)));
+
+        let input = "\0,tag1=1,tag2=2 value=1 123";
+        let vals = parse(input);
+        assert!(matches!(vals, Err(Error::MeasurementValueInvalid)));
```
Comment (alamb) [inline influxdb_line_protocol/src/lib.rs:1287]: I suggest adding tests for null bytes appearing in the tag names, tag values and field names and field values

Perhaps these cases
```
m,tag\01=one,tag2=2 value=1 123
m,tag1=o\0ne,tag2=2 value=1 123
m,tag1=one,tag2=2 val\0ue=1 123
m,tag1=one,tag2=2 value="v\0" 123
```

File: influxdb_line_protocol/src/builder.rs (hunk 1 above)
Comment (alamb) [reply influxdb_line_protocol/src/builder.rs:421]: Did you compare this behavior with the go based line protocol parser (aka what does `influxdb` support)? Maybe @stuartcarnie  can give us a note or suggested way to test.

Logically it makes a lot of sense to me to not allow `=` in the measurement names, but if the golang parser allows it I think we should also continue allowing such input in the Rust parser

**Review by alamb — COMMENTED** [review]
Thank you for your work on this issue @wiedld 

The code seems to do what it says it is supposed to do 👍 

However, I am not 100% sure if this is the right behavior -- can you please verify that this new behavior matches the existing golang parser (aka what influxdb) supports? If not, I think we should make sure we deliberately make the choice to not be compatible and document that choice somewhere.

File: influxdb_line_protocol/src/builder.rs (hunk 1 above)
Comment (wiedld) [reply influxdb_line_protocol/src/builder.rs:421]: Added some test cases [on a branch in influxdb](https://github.com/influxdata/influxdb/commit/69aa859810be6e32b6ddec46e4768c04c9404123). The parser itself (outside of any other bits) has:
* the same handling of nulls as this PR.
* diverges on the eq-within-measurement. That pattern is accepted by the parser.

Will update the iox parser to match.

Comment (stuartcarnie) [reply influxdb_line_protocol/src/builder.rs:421]: @wiedld and I paired up to look and confirmed `=` is an acceptable character for measurement names in InfluxDB OG:

```
> create database meas
> use meas
Using database meas
> insert measurement=withEq,tag0=v0 value=1
> show measurements
name: measurements
name
----
measurement=withEq
> select * from "measurement=withEq"
name: measurement=withEq
time                        tag0 value
----                        ---- -----
2023-05-16T00:56:45.778733Z v0   1
```

File: influxdb_iox/tests/query_tests2/cases.rs [hunk 5]
```diff
@@ -169,6 +169,18 @@ async fn periods() {
     .await;
 }
 
+#[tokio::test]
+async fn equals() {
+    test_helpers::maybe_start_logging();
+
+    TestCase {
+        input: "cases/in/equals.sql",
```
Comment (wiedld) [inline influxdb_iox/tests/query_tests2/cases.rs:177]: end_to_end test case per excellent suggestion by @stuartcarnie

**Review by alamb — APPROVED** [review]
I think this looks great @wiedld  🏅

**Pull request merged**